lib/core/macros.nim

```diff
@@ -1223,33 +1223,58 @@ macro expandMacros*(body: typed): untyped =
   echo result.toStrLit
 
 proc customPragmaNode(n: NimNode): NimNode =
-  expectKind(n, {nnkSym, nnkDotExpr})
-  if n.kind == nnkSym:
-    let sym = n.getImpl()
-    sym.expectRoutine()
-    result = sym.pragma
-  elif n.kind == nnkDotExpr:
-    let typDef = getImpl(getTypeInst(n[0]))
-    typDef.expectKind(nnkTypeDef)
-    typDef[2].expectKind(nnkObjectTy)
-    let recList = typDef[2][2]
-    for identDefs in recList:
-      for i in 0 .. identDefs.len - 3:
-        if identDefs[i].kind == nnkPragmaExpr and
-           identDefs[i][0].kind == nnkIdent and $identDefs[i][0] == $n[1]:
-          return identDefs[i][1]
+  expectKind(n, {nnkSym, nnkDotExpr, nnkBracketExpr, nnkTypeOfExpr})
+  let 
+    typ = n.getTypeInst()
+
+  if n.kind == nnkSym: # either an variable or a proc
+    let impl = n.getImpl()
+    if impl.kind in RoutineNodes:
+      return impl.pragma
+
+  if typ.typeKind == ntyTypeDesc or n.kind == nnkSym:
+    let impl = (if typ.typeKind == ntyTypeDesc: typ[1] else: typ).getImpl()
+    return impl[0][1]
+  if n.kind == nnkDotExpr:
+    var typDef = getImpl(getTypeInst((if n[0].kind == nnkHiddenDeref: n[0][0] else: n[0])).symbol)
+    while typDef != nil:
+      typDef.expectKind(nnkTypeDef)
+      typDef[2].expectKind({nnkRefTy, nnkPtrTy, nnkObjectTy})
+      let isRef = typDef[2].kind in {nnkRefTy, nnkPtrTy}
+      if isRef and typDef[2][0].kind == nnkSym: # defines ref type for another object
```

Comment:
what will happen if  ``typDef[2][0].kind`` is not nnkSym? 
Sounds like you need an expectKind instead

Reply:
If the type is a `ref` type, `typDef[2][0]` is either a sym in cases like `NodeRef = ref Node` or of the kind `nnkObjectTy` if the object definition is embedded. But now as you mention it, I also add `nnkBracketExpr` in case one is inheriting from a generic